src/Symfony/Component/Workflow/Dumper/MermaidDumper.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Workflow\Dumper;
+
+use Symfony\Component\Workflow\Definition;
+use Symfony\Component\Workflow\Exception\InvalidArgumentException;
+use Symfony\Component\Workflow\Marking;
+
+class MermaidDumper implements DumperInterface
+{
+    public const DIRECTION_TOP_TO_BOTTOM = 'TB';
+    public const DIRECTION_TOP_DOWN = 'TD';
+    public const DIRECTION_BOTTOM_TO_TOP = 'BT';
+    public const DIRECTION_RIGHT_TO_LEFT = 'RL';
+    public const DIRECTION_LEFT_TO_RIGHT = 'LR';
+
+    private const VALID_DIRECTIONS = [
+        self::DIRECTION_TOP_TO_BOTTOM,
+        self::DIRECTION_TOP_DOWN,
+        self::DIRECTION_BOTTOM_TO_TOP,
+        self::DIRECTION_RIGHT_TO_LEFT,
+        self::DIRECTION_LEFT_TO_RIGHT,
+    ];
+
+    /**
+     * @var string
+     */
+    private $direction;
+
+    public function __construct(string $direction = self::DIRECTION_TOP_DOWN)
+    {
+        if (!\in_array($direction, self::VALID_DIRECTIONS, true)) {
+            throw new InvalidArgumentException("Direction '{$direction}' is not valid.");
+        }
+
+        $this->direction = $direction;
+    }
+
+    public function dump(Definition $definition, Marking $marking = null, array $options = []): string
+    {
+        $output = ["graph {$this->direction}"];
+
+        $meta = $definition->getMetadataStore();
+
+        foreach ($definition->getPlaces() as $place) {
+            $placeInfo = [
+                'name' => $place,
+                'meta' => $meta->getPlaceMetadata($place),
+                'isInitial' => \in_array($place, $definition->getInitialPlaces()),
+            ];
+
+            $output[] = $this->formatPlace($placeInfo);
+        }
+
+        foreach ($definition->getTransitions() as $transition) {
+            foreach ($transition->getFroms() as $from) {
+                foreach ($transition->getTos() as $to) {
+                    $output[] = "{$from}-->|{$transition->getName()}|{$to}";
+                }
+            }
+        }
+
+        return implode(\PHP_EOL, $output);
+    }
+
+    protected function formatPlace(array $placeInfo): string
+    {
+        $placeName = $placeInfo['name'];
+
+        $placeLabel = $placeName;
+        if (\array_key_exists('label', $placeInfo['meta'])) {
+            $placeLabel = $placeInfo['meta']['label'];
+        }
+
+        $labelShape = '[%s]';
+        if ($placeInfo['isInitial']) {
+            $labelShape = '([%s])';
+        }
+
+        $placeNodeFormat = '%s'.$labelShape;
+        $placeNode = sprintf($placeNodeFormat, $placeName, $placeLabel);
```

Comment:
you need to apply proper escaping on labels, to ensure the generated graph is valid, because labels can be anything: https://mermaid-js.github.io/mermaid/#/flowchart?id=special-characters-that-break-syntax

Reply:
also, be careful than the lowercase `end` identifier is reserved and cannot be used as a node identifier.